src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs

```diff
@@ -131,6 +125,56 @@ public CompilationResult Compile([NotNull] RelativeFileInfo fileInfo, [NotNull]
             }
         }
 
+        // Internal for unit testing
+        internal CompilationResult GetCompilationFailedResult(
+            string relativePath,
+            string compilationContent,
+            string assemblyName,
+            IEnumerable<Diagnostic> diagnostics)
+        {
+            var diagnosticGroups = diagnostics
+                .Where(IsError)
+                .GroupBy(diagnostic => GetFilePath(relativePath, diagnostic), StringComparer.Ordinal);
+
+            var failures = new List<ICompilationFailure>();
+            foreach (var group in diagnosticGroups)
+            {
+                var sourceFilePath = group.Key;
+                string sourceFileContent;
+                if (string.Equals(assemblyName, sourceFilePath, StringComparison.Ordinal))
```

Comment:
This allows us to show errors if they happen to be produced in the generated code (for instance a custom base type that does have a `RenderAsync` method). Mapping it to the cshtml file is not very helpful in diagnosing the issue.